### PR TITLE
[Data] [Train] Add caching variants to multi-node train benchmark

### LINKED ISSUE #38282: [Data] Multi node train benchmark yields lower throughput with more workers added
### What happened + What you expected to happen

For the release tests read_images_train_4_gpu and read_images_train_16_gpu, the main logic is to ingest images into a TorchTrainer with 4 or 16 workers. We would expect that throughput should be higher for the 16 worker case than the 4 worker case, but we observed the inverse when running the tests on [Buildkite](https://buildkite.com/ray-project/release-tests-pr/builds/48268#0189d9b1-f2cd-47e0-9703-15bf4b3f42b9/383-405):

The 4 worker case had a tput of 58.17 img/s, while the 16 worker case had a tput of 39.22 img/s.

### Versions / Dependencies

ray master
python 3.9

### Reproduction script

read_images_train_4_gpu, read_images_train_16_gpu

### Issue Severity

None

### PR DESCRIPTION
**Pull request by scottjlee**
## Why are these changes needed?
- Generalize Benchmark utility to support running a function, instead of only executing a dataset.
- Add full support for multi-node, GPU case with configurable flags
- Add benchmark variants involving caching the dataset at different points

## Related issue number
Closes #38282

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: release/nightly_tests/dataset/multi_node_train_benchmark.py [hunk 1]
```diff
@@ -31,6 +40,18 @@ def parse_args():
...
     )
+    parser.add_argument(
+        "--repeat-ds",
+        default=1,
+        type=int,
+        help="Read the input dataset n times, used to increase the total data size."
+    )
+    parser.add_argument(
+        "--read-task-cpus",
+        default=2,
```
Comment (stephanie-wang) [inline release/nightly_tests/dataset/multi_node_train_benchmark.py:51]: Probably make this 1 by default? Or just remove, I'm not sure we need it, although it'd be good to confirm that changing this parameter doesn't affect the final throughput if the output dataset is cached.

File: release/nightly_tests/dataset/multi_node_train_benchmark.py [hunk 2]
```diff
@@ -66,9 +87,10 @@ def parse_args():
     args = parser.parse_args()
 
     if args.file_type == "image":
-        args.data_root = "s3://air-cuj-imagenet-1gb"
+        args.data_root = "s3://anonymous@air-example-data-2/20G-image-data-synthetic-raw"
```
Comment (stephanie-wang) [inline release/nightly_tests/dataset/multi_node_train_benchmark.py:90]: Ideally add this as a parsable argument.

File: release/nightly_tests/dataset/multi_node_train_benchmark.py [hunk 3]
```diff
@@ -108,71 +130,128 @@ def crop_and_flip_image_batch(image_batch):
...
+                datasets: Dict[str, Dataset],
+                world_size: int,
+                worker_handles: Optional[List[ActorHandle]],
+                worker_node_ids: Optional[List[NodeIdStr]],
+                **kwargs,
+            ) -> List[Dict[str, DataIterator]]:
+                assert len(datasets) == 1, "This example only handles the simple case"
+
+                # Split the dataset into shards, materializing the results.
+                materialized_shards = datasets["train"].split(
```
Comment (stephanie-wang) [inline release/nightly_tests/dataset/multi_node_train_benchmark.py:214]: Hmm one issue here is that we actually want to partition the input dataset, not the output one. But let's focus on stabilizing the other benchmarks first, since they are more similar to each other than this one. We can follow up on the pre-partitioning in a separate PR.

File: release/nightly_tests/dataset/multi_node_train_benchmark.py [hunk 4]
```diff
@@ -108,71 +130,128 @@ def crop_and_flip_image_batch(image_batch):
...
-    for label, tput in metrics.items():
-        metrics_dict[label].update({"THROUGHPUT": tput})
-
-    test_name = f"read_{args.file_type}_train_{args.num_workers}workers"
-    result_dict = {
-        test_name: metrics_dict,
-        "success": 1,
-    }
+    # Report the throughput of one epoch (averaged across epochs)
+    runtime_one_epoch = (end_t - start_t) / args.num_epochs
```
Comment (stephanie-wang) [inline release/nightly_tests/dataset/multi_node_train_benchmark.py:240]: We should drop the first epoch to avoid counting warmup time.

File: release/nightly_tests/dataset/multi_node_train_benchmark.py [hunk 5]
```diff
@@ -108,71 +130,128 @@ def crop_and_flip_image_batch(image_batch):
...
         raise Exception(f"Unknown file type {args.file_type}")
+    
+    if args.repeat_ds > 1:
+        ray_dataset = ray_dataset.union(*[Dataset.copy(ray_dataset) for _ in range(args.repeat_ds-1)])
+
+    if cache_input_ds:
+        ray_dataset = ray_dataset.materialize()
+
     # 2) Preprocess data by applying transformation with map_batches()
     ray_dataset = ray_dataset.map_batches(crop_and_flip_image_batch)
```
Comment (stephanie-wang) [inline release/nightly_tests/dataset/multi_node_train_benchmark.py:170]: Here we should pass args.batch_size.

**Review by stephanie-wang — COMMENTED** [review]
Other notes:
- we can use CPU-only nodes for this benchmark
- probably we can reduce the number of epochs too. 3 is fine to get a tput reading
- it'd be good to report the bytes spilled/restored. You can check out python/ray/tests/test_object_spilling.py for some sample code on how to read these metrics.

File: release/nightly_tests/dataset/multi_node_train_benchmark.py (hunk 1 above)
Comment (scottjlee) [reply release/nightly_tests/dataset/multi_node_train_benchmark.py:51]: i intended to use this parameter for the case where we want to test caching with extra CPU nodes, where we can specify the number of CPUs to actually use -- do you think instead of doing that, i should just run it as a separate release test with a different compute config?

File: release/nightly_tests/dataset/multi_node_train_benchmark.py [hunk 6]
```diff
@@ -98,81 +146,212 @@ def get_transform(to_torch_tensor):
...
-        train_loop_per_worker,
-        datasets={"train": ray_dataset},
-        scaling_config=ScalingConfig(num_workers=args.num_workers, use_gpu=True),
-        dataset_config=ray.train.DataConfig(
-            execution_options=options,
-        ),
-    )
+    if prepartition_ds:
+
+        class PrepartitionCacheDataConfig(DataConfig):
```
Comment (stephanie-wang) [inline release/nightly_tests/dataset/multi_node_train_benchmark.py:239]: Shall we just remove this as it's not used?

File: release/nightly_tests/dataset/multi_node_train_benchmark.py [hunk 7]
```diff
@@ -98,81 +146,212 @@ def get_transform(to_torch_tensor):
...
+            ),
+        )
+    else:
+        torch_trainer = TorchTrainer(
+            train_loop_per_worker,
+            datasets={"train": ray_dataset},
+            scaling_config=ScalingConfig(
+                num_workers=args.num_workers - 1,
+                use_gpu=args.use_gpu,
+                placement_strategy="STRICT_SPREAD",
```
Comment (stephanie-wang) [inline release/nightly_tests/dataset/multi_node_train_benchmark.py:292]: Can you add a note about these params?

**Review by stephanie-wang — APPROVED** [review]

File: release/nightly_tests/dataset/multi_node_train_benchmark.py (hunk 6 above)
Comment (scottjlee) [reply release/nightly_tests/dataset/multi_node_train_benchmark.py:239]: yeah good point, we can re-implement this correctly in a future PR.

File: release/nightly_tests/dataset/multi_node_train_benchmark.py [hunk 8]
```diff
@@ -98,81 +146,212 @@ def get_transform(to_torch_tensor):
...
                 batch_size=args.batch_size,
-                local_shuffle_buffer_size=args.local_shuffle_buffer_size,
-                prefetch_batches=10,
             ):
                 num_rows += args.batch_size
             end_t = time.time()
-            # Record throughput per epoch.
-            epoch_tput = num_rows / (end_t - start_t)
-            train.report({"tput": epoch_tput, "epoch": i})
+        # Workaround to report the final epoch time from each worker, so that we
```
Comment (raulchen) [inline release/nightly_tests/dataset/multi_node_train_benchmark.py:205]: nit: move this loop function out of the main function? as it's getting too long.
